### PR TITLE
Pass custom UA with client.__init__

### PR DESCRIPTION
**Pull request by vshlapakov**
Review, please.

### CODE REVIEW

File: hubstorage/client.py [hunk 1]
```diff
@@ -49,7 +49,8 @@ class HubstorageClient(object):
     RETRY_DEFAULT_JITTER_MS = 500
     RETRY_DEFAULT_EXPONENTIAL_BACKOFF_MS = 500
 
-    def __init__(self, auth=None, endpoint=None, connection_timeout=None, max_retries=None, max_retry_time=None):
+    def __init__(self, auth=None, endpoint=None, connection_timeout=None,
+                 max_retries=None, max_retry_time=None, useragent=None):
```
Comment (bertinatto) [inline hubstorage/client.py]: Just a note (implement it if you want :smile:  ): perhaps it'd be better if `useragent` was named `user_agent`, to follow the notation used in the other arguments.

File: tests/test_client.py [hunk 2]
```diff
@@ -2,10 +2,17 @@
...
+from hubstorage import HubstorageClient
 from hubstorage.utils import millitime, apipoll
 
 class ClientTest(HSTestCase):
 
+    def test_custom_ua(self):
+        c = HubstorageClient(auth=cls.auth,
+                             endpoint=cls.endpoint,
+                             useragent='testUA')
+        self.assertEqual(c.useragent == 'testUA')
```
Comment (bertinatto) [inline tests/test_client.py]: Maybe another assertion to cover the case where the user doesn't pass in a custom value for `useragent`?

**Comment by bertinatto** [comment]
LGTM, just added two suggestions.

File: hubstorage/client.py (hunk 1 above)
Comment (vshlapakov) [inline hubstorage/client.py]: I had doubts about it, but your vote overcame it, fixed.

File: hubstorage/client.py [hunk 3]
```diff
@@ -40,7 +40,7 @@ def _hc_retry_on_exception(err):
 class HubstorageClient(object):
 
     DEFAULT_ENDPOINT = 'https://storage.scrapinghub.com/'
-    USERAGENT = 'python-hubstorage/{0}'.format(__version__)
+    DEFAULT_USER_AGENT = 'python-hubstorage/{0}'.format(__version__)
```
Comment (chekunkov) [inline hubstorage/client.py]: this breaks backwards compatibility

Comment (chekunkov) [inline hubstorage/client.py]: You can keep `DEFAULT_USER_AGENT`, but object should as well contain `USERAGENT` attribute - e.g. as property that returns `DEFAULT_USER_AGENT` with deprecation warning and appropriate setter

Comment (vshlapakov) [inline hubstorage/client.py]: Make sense, I didn't realise that keeping the attribute name is important but you're right, redefining it should be a common practice. Thanks!

@chekunkov  Have just fixed it, let me know if we miss anything else here.

Comment (chekunkov) [inline hubstorage/client.py]: @vshlapakov now I'm afraid that property approach not going to work. I've found 2 examples in our code base that don't work with it:

1)

```
In [8]: class HubstorageClientWithCustomUA(HubstorageClient):
   ...:     USERAGENT = 'b'
   ...:

In [9]: hsc = HubstorageClientWithCustomUA()

In [10]: hsc.USERAGENT
Out[10]: 'b'

In [11]: hsc.DEFAULT_USER_AGENT
Out[11]: 'python-hubstorage/0.23.1'

In [12]: hsc.user_agent
Out[12]: 'python-hubstorage/0.23.1'
```

2) 

```
In [13]: class HubstorageClientWithCustomUA(HubstorageClient):
    USERAGENT = HubstorageClient.USERAGENT + "/c"
   ....:
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-13-df6deb570a9f> in <module>()
----> 1 class HubstorageClientWithCustomUA(HubstorageClient):
      2     USERAGENT = HubstorageClient.USERAGENT + "/c"
      3

<ipython-input-13-df6deb570a9f> in HubstorageClientWithCustomUA()
      1 class HubstorageClientWithCustomUA(HubstorageClient):
----> 2     USERAGENT = HubstorageClient.USERAGENT + "/c"
      3

TypeError: unsupported operand type(s) for +: 'property' and 'str'
```

--

Instead we can set `USERAGENT = None` and then in `_create_session`:

```
        s.headers.update({'User-Agent': self.USERAGENT or self.user_agent})
```

and raise deprecation warning if `USERAGENT` is not `None`

File: hubstorage/client.py [hunk 4]
```diff
@@ -162,6 +170,16 @@ def close(self, timeout=None):
         if self._batchuploader is not None:
             self.batchuploader.close(timeout)
 
+    @property
+    def USERAGENT(self):
+        warnings.warn(USERAGENT_DEPRECATED_MSG, DeprecationWarning)
```
Comment (chekunkov) [inline hubstorage/client.py]: `warnings` is not imported, I have `NameError: global name 'warnings' is not defined` locally and wonder how tests passed

Comment (vshlapakov) [inline hubstorage/client.py]: @chekunkov Looks like tests are disabled for pretty long period of time:
https://github.com/scrapinghub/python-hubstorage/commit/21f520b7bbc411eedbca9284bde0d57d72b46ffe
@dangra I guess we can restore it now?

File: hubstorage/client.py (hunk 3 above)
Comment (vshlapakov) [inline hubstorage/client.py]: If `HubstorageClient.USERAGENT + "/c"` is not an artificial case, then `USERAGENT = None` won't work, I can set it to an empty string instead.

Comment (chekunkov) [inline hubstorage/client.py]: it's not an artificial case, you're right

Comment (chekunkov) [inline hubstorage/client.py]: but then it won't work as expected, resulting in `USERAGENT == '/c'`. Maybe then we should default to `USERAGENT = DEFAULT_USER_AGENT` and then raise warning if they are not equal?

Comment (vshlapakov) [inline hubstorage/client.py]: Agree, it's better. We still don't raise a warning when somebody reads `cls.USERAGENT` (like we did it via property), but I guess it's an acceptable trade-off and it's fine to just ignore such cases for now.

File: hubstorage/client.py [hunk 5]
```diff
@@ -135,7 +146,9 @@ def _create_retrier(self, max_retries, max_retry_time):
 
     def _create_session(self):
         s = session()
-        s.headers.update({'User-Agent': self.USERAGENT})
+        s.headers.update({'User-Agent': self.USERAGENT or self.user_agent})
```
Comment (chekunkov) [inline hubstorage/client.py]: if `self.USERAGENT` is not empty (and we agreed that `USERAGENT = DEFAULT_USER_AGENT`) `self.user_agent` will not be used.

File: hubstorage/client.py (hunk 3 above)
Comment (chekunkov) [inline hubstorage/client.py]: > We still don't raise a warning when somebody reads cls.USERAGENT (like we did it via property), but I guess it's an acceptable trade-off and it's fine to just ignore such cases for now.

@redapple do you have any good suggestion on how to provide backwards compatible read with warning here?

File: hubstorage/client.py (hunk 5 above)
Comment (vshlapakov) [inline hubstorage/client.py]: Good point, updated

File: hubstorage/client.py (hunk 4 above)
Comment (vshlapakov) [inline hubstorage/client.py]: Looks like current tests were created around older testing environment which is outdated now and doesn't work as is even after a few fixes, I propose to recover tests in a separate PR.

**Comment by redapple** [comment]
I think I'm missing context a bit.
Why do you want to change `USER_AGENT` to have `DEFAULT_USER_AGENT`? (honest question, no judgement)
Is it to make it more explicit that it can be overriden?
Another way to ask the question, if you keep `USER_AGENT` as it is now, is it compatible with the ` ..., user_agent=None)` change?
I'm not sure many users outside Scrapinghub customize the User-Agent (do you have some numbers from the logs?), so it could be that it only needs to be clear in the docs that one should use `user_agent` arg at init and make sure this is fixed in Scrapinghub projects.

Bottom line: changing class attributes names and behavior is a real pain: see https://github.com/scrapy/scrapy/pull/1891 up to https://github.com/scrapy/scrapy/pull/2243

**Comment by chekunkov** [comment]
@redapple good point, from logs it seems that customers either use python-hubstorage with default user agent or use some other libraries like requests or RestSharp. I'm pretty sure that covers 99% cases. The only cases where USERAGENT is modified I've found are internal, versions are pinned and we can update them with appropriate changes. Do you agree it doesn't worth an effort to bother with backwards compatibility in this case? We can just keep it simple then - just how it was implemented initially with `DEFAULT_USER_AGENT` and `self.user_agent = user_agent or self.USERAGENT` in `__init__`.

> Why do you want to change USER_AGENT to have DEFAULT_USER_AGENT? (honest question, no judgement)
> Is it to make it more explicit that it can be overriden?

I think so, maybe also to keep it the same as `DEFAULT_ENDPOINT` and `DEFAULT_CONNECTION_TIMEOUT_S` /cc @vshlapakov

**Comment by redapple** [comment]
Right, it does not look like worth the effort. Explicit enough release notes should suffice.

**Comment by vshlapakov** [comment]
Thanks for the comments, updated :white_check_mark:
